### PR TITLE
[Bug fix] LLK: preserve UInt16 datums whose low byte is zero across SrcA/SrcB reconfig (#37571)

### LINKED ISSUE #37571: ttnn.sort returns duplicate indices (not a valid permutation) for tensors wider than 256 elements (VLLM TopK Sampling)
Not a regression, but found when debugging garbage outputs through VLLM with TopK sampling using recent tt-metal 430d1f4d95d8401df010d0d932807258e8605ab1 from Feb 7 - maybe LLK related, but not sure to tag label for now. See details below, I thank Claude Code for doing most of the debug here :)

## Describe the bug

`ttnn.sort` returns indices that are **not a valid permutation** when sorting tensors with more than 256 elements (Wt > 8 tiles) along the last dimension. The sorted **values** are correct and monotonically ordered, but the returned **indices** contain duplicates and missing values.

Specifically:
- **Index 0** is always duplicated (appears 2x for Wt=16, 4x for Wt=32, 8x for Wt=64, etc.)
- **Multiples of 256** are always missing (index 256, 512, 768, ...)
- The bug is **data-independent** — occurs for random values, all-zeros, pre-sorted ascending, and pre-sorted descending inputs
- The bug affects **both ascending and descending** sort directions

This breaks any code that relies on sort indices for scatter/gather operations. In our case, it causes vLLM's top-k/top-p sampler (`apply_top_k_top_p`) to produce corrupted logits, resulting in garbage text output during non-greedy sampling.

### Minimal repro

```python
import torch, ttnn

device = ttnn.open_device(device_id=0)

t = torch.randn([1, 512], dtype=torch.bfloat16)
inp = ttnn.from_torch(t, ttnn.bfloat16, layout=ttnn.Layout.TILE, device=device)
_, indices = ttnn.sort(inp, dim=-1)
idx = ttnn.to_torch(indices)[0].to(torch.int64).tolist()

print(f"Unique indices: {len(set(idx))}/512")   # 511/512
print(f"Index 0 count:   {idx.count(0)}")        # 2  (expected 1)
print(f"Index 256 count: {idx.count(256)}")       # 0  (expected 1)

ttnn.close_device(device)
```

Output:
```
Unique indices: 511/512
Index 0 count:   2  (expected 1)
Index 256 count: 0  (expected 1)
```

### Failure boundary

| Wt (tiles) | Elements | Status | Index 0 dup count | Missing indices |
|------------|----------|--------|--------------------|-----------------|
| 8 | 256 | PASS | 1 (correct) | none |
| 9 | 288 | **FAIL** | 2 | [256] |
| 16 | 512 | **FAIL** | 2 | [256] |
| 32 | 1024 | **FAIL** | 4 | [256, 512, 768] |
| 64 | 2048 | **FAIL** | 8 | [256, 512, ..., 1792] |

The bug first appears when the bitonic sort requires a **4th merge stage** (Wt=9 pads to 16, needing `log2(16)=4` stages vs `log2(8)=3` stages for Wt=8).

## Steps to Reproduce

1. Run any of the repro scripts from this gist: https://gist.github.com/kmabeeTT/d5d75c62219e21c44cd386b74eb02736
   - `test_ttnn_sort_minimal.py` — 15-line minimal repro shown above
   - `test_ttnn_sort_repro.py` — comprehensive sweep across sizes (32 to 151936)
   - `test_ttnn_sort_boundary.py` — pinpoints exact boundary, tests data independence

2. Or inline:
```bash
python3 test_ttnn_sort_minimal.py
```

## Expected Behavior

`ttnn.sort` should return indices that form a valid permutation — each index from `0` to `N-1` should appear exactly once. This is required for sort+scatter roundtrips to preserve data, and for `torch.gather(input, dim, sort_indices)` to reconstruct the sorted values.

### Existing test gap

The current `test_sort_indices` test `tests/ttnn/unit_tests/operations/data_movement/test_sort.py` only covers shapes where the sort dimension is ≤ 64 elements (Wt ≤ 2), which is below the failure threshold of Wt > 8. The bug is not caught by existing tests.

## Additional context

- Tested on Wormhole B0
- bf16 input, TILE layout
- Affects all three program factory paths (single-core for Wt ≤ 64, cross-core and multi-core DRAM for larger)
- The sorted values are always correct — only the indices are wrong
- Root cause appears to be in the LLK bitonic merge (`_bitonic_topk_merge` in `ckernel_sfpu_topk.h`), specifically in how index load/store addresses are calculated when the merge crosses the 8-tile (256-element) boundary

### PR DESCRIPTION
**Pull request by nstamatovicTT**
### PR Category

Bug fix

### Summary

Fixes #37571.

`ttnn.sort` returned non-permutation indices for last-dim > 256 elements on Wormhole B0 (and Blackhole). The corruption was data-independent: index `0` duplicated, multiples of 256 (`256, 512, 768, …`) missing.

**Root cause** is in the LLK unpacker reconfig path:

1. `configure_unpack_AB` (called from the first `_llk_unpack_hw_configure_`) explicitly sets `ALU_ACC_CTRL_Zero_Flag_disabled_src` based on the *initial* dst format.
2. `ttnn.sort`'s compute kernel boots with bf16 (values), so the flag is set to **0** (flush ON).
3. The kernel later reconfigs SrcA to UInt16 to handle index tiles via `_llk_unpack_reconfig_data_format_srca_impl_`. That path **never updated the zero-flag**.
4. With the flag still at 0, `MOVA2D` (used by the math TRISC's datacopy) applies its bf16-denormal-flush rule `(SrcAVal & 0xFF) == 0 → 0` to UInt16 *index* data, eating any value with a zero low byte.

The LLK already documents this concern in `llk_unpack_common.h:173-191` and ships the helper `_llk_unpack_reconfig_zero_src_flag_`, but it was only wired into `configure_unpack_AB` — not into the runtime SrcA/SrcB reconfig functions. This PR closes that gap.

**Fix:** at the end of `_llk_unpack_reconfig_data_format_srca_impl_` and `_srcb_impl_` (both WH B0 and BH), re-assert `ALU_ACC_CTRL_Zero_Flag_disabled_src = 1` whenever the new dst format is `UInt16`. The write is sticky and only sets to 1 — never re-enables the flag — matching `_llk_unpack_reconfig_zero_src_flag_`'s documented semantics.

**Verification (Wormhole B0):**

- Issue's verbatim minimal repro: `Unique indices: 512/512`, `Index 0 count: 1`, `Index 256 count: 1` — was `511/512`, `2`, `0` before fix.
- All four scenarios from the issue's table pass: widths 256, 288, 512, 1024 — every index in `range(width)` appears exactly once.
- New LLK regression test `tests/sources/sort_test.cpp` + `tests/python_tests/test_sort.py` reproduces the exact #37571 signature pre-fix (Wt=8 PASS, Wt=16 missing `[256]`, Wt=32 missing `[256, 512, 768]`) and passes post-fix.

### Notes for reviewers

- Behavior change is conditional and monotonic: only writes when `dst_format == UInt16`, only ever writes `1`. For all other format reconfigs the change is a pure no-op (one extra compare + not-taken branch in a cold init function).

### CI Status
_Auto-generated on every push. Badges update live. Click a badge to filter runs by this branch._

- [](https://github.com/tenstorrent/tt-metal/actions/workflows/sanity-tests.yaml?query=branch:nstamatovic/issue-37571-fix-uint16-zero-flag)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/runtime-sanity-tests.yaml?query=branch:nstamatovic/issue-37571-fix-uint16-zero-flag)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/blackhole-post-commit.yaml?query=branch:nstamatovic/issue-37571-fix-uint16-zero-flag)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/tt-metal-l2-nightly.yaml?query=branch:nstamatovic/issue-37571-fix-uint16-zero-flag)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/all-model-tests.yaml?query=branch:nstamatovic/issue-37571-fix-uint16-zero-flag)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select.yaml?query=branch:nstamatovic/issue-37571-fix-uint16-zero-flag)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-t3k.yaml?query=branch:nstamatovic/issue-37571-fix-uint16-zero-flag)
- [](https://github.com/tenstorrent/tt-metal/actions/workflows/pipeline-select-galaxy.yaml?query=branch:nstamatovic/issue-37571-fix-uint16-zero-flag)

### CODE REVIEW

File: tt_metal/tt-llk/tests/python_tests/test_sort.py [hunk 1]
```diff
@@ -0,0 +1,268 @@
+# SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
```
Comment (pmilenkovicTT) [inline tt_metal/tt-llk/tests/python_tests/test_sort.py]: 2026

File: tt_metal/tt-llk/tests/sources/sort_test.cpp [hunk 2]
```diff
@@ -0,0 +1,489 @@
+// SPDX-FileCopyrightText: © 2025 Tenstorrent AI ULC
```
Comment (pmilenkovicTT) [inline tt_metal/tt-llk/tests/sources/sort_test.cpp]: 2026

File: tt_metal/tt-llk/tests/sources/sort_test.cpp [hunk 3]
```diff
@@ -0,0 +1,489 @@
...
+#include "llk_lib_pack_wrappers.h"
+#include "llk_pack.h"
+#include "llk_pack_common.h"
+
+void run_kernel(RUNTIME_PARAMETERS params)
+{
+#if defined(RUNTIME_FORMATS) && !defined(SPEED_OF_LIGHT)
+    const FormatConfig& formats = params.formats;
+#endif
+    const int FULL_CT_DIM = params.FULL_CT_DIM;
```
Comment (pmilenkovicTT) [inline tt_metal/tt-llk/tests/sources/sort_test.cpp]: A bit cosmetic, but should we be mixing `int` and `std::uint32_t`?

File: tt_metal/tt-llk/tests/sources/sort_test.cpp [hunk 4]
```diff
@@ -0,0 +1,489 @@
...
+        else
+        {
+#ifdef ARCH_BLACKHOLE
+            _llk_pack_reconfig_data_format_<is_fp32_dest_acc_en, false>(pack_src_format, pack_dst_format, 16 * 16 * 4, FACE_R_DIM, TILE_C_DIM, 4, false, 1);
+#else
+            _llk_pack_reconfig_data_format_<is_fp32_dest_acc_en, false>(pack_src_format, pack_dst_format, 16 * 16 * 4, FACE_R_DIM, 4, false, false);
+#endif
+        }
+
+        _llk_pack_init_wrapper_<false, false>(pack_dst_format);
```
Comment (pmilenkovicTT) [inline tt_metal/tt-llk/tests/sources/sort_test.cpp]: Can we mark what each parameter is here?

**Review by pmilenkovicTT — COMMENTED** [review]
Looks good! A few (mostly cosmetic) notes:

File: tt_metal/tt-llk/tests/sources/sort_test.cpp [hunk 5]
```diff
@@ -0,0 +1,489 @@
...
+        {
+            _llk_unpack_reconfig_data_format_srca_impl_<is_fp32_dest_acc_en, p_dim_stride_target::IGNORE, false>(
+                unpack_src_format, unpack_dst_format, 16 * 16 * 4);
+        }
+
+        _llk_unpack_A_init_<BroadcastType::NONE, false, EltwiseBinaryReuseDestType::NONE, unpack_to_dest>(
+            transpose_faces ? 1 : 0, transpose_faces ? 1 : 0, FACE_R_DIM, 4, unpack_src_format, unpack_dst_format);
+    };
+
+    auto unpack_tile = [&](Stage stage, int l1_tile_index)
```
Comment (skotaracTT) [inline tt_metal/tt-llk/tests/sources/sort_test.cpp]: Even though it does make for cleaner looking code, these functions don't add much complexity other than being thin wrappers. Visually confusing, function calls resemble compute api.

In LLK test infra, we strip wrappers and simply call llk primitives directly. Might be good to keep test files consistent for readability.

**Review by skotaracTT — APPROVED** [review]

File: tt_metal/tt-llk/tests/python_tests/test_sort.py [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+            [(v, c) for v, c in idx_counter.items() if c > 1], key=lambda x: x[0]
+        )
+
+        if missing or duplicates:
+            all_ok = False
+            if first_failure_row is None:
+                first_failure_row = row_idx
+                missing_mod_256 = [v for v in missing if v % 256 == 0]
+                dup0_count = idx_counter.get(0, 0)
+                print(
```
Comment (fvranicTT) [inline tt_metal/tt-llk/tests/python_tests/test_sort.py]: Please don't use raw prints. Either remove this or use the logger. Thanks.

File: tt_metal/tt-llk/tt_llk_wormhole_b0/llk_lib/llk_unpack_common.h [hunk 7]
```diff
@@ -123,6 +123,12 @@ inline void _llk_unpack_reconfig_data_format_srca_impl_(
 
         TT_SETADCXX(p_setadc::UNP_A, (unpack_face_r_dim << 4) - 1, 0x0);
     }
+
+    // tt-metal#37571: don't let MOVA2D flush UInt16 (256, 512, …) as bf16 -0.
```
Comment (ncvetkovicTT) [inline tt_metal/tt-llk/tt_llk_wormhole_b0/llk_lib/llk_unpack_common.h]: MOVA2D only in this case, but what about regular L1->SrcReg->FPU->DEST path?

File: tt_metal/tt-llk/tt_llk_wormhole_b0/llk_lib/llk_unpack_common.h [hunk 8]
```diff
@@ -168,6 +174,12 @@ inline void _llk_unpack_reconfig_data_format_srcb_impl_(
 
         TT_SETADCXX(p_setadc::UNP_B, (unpack_face_r_dim << 4) - 1, 0x0);
     }
+
+    // tt-metal#37571: see _llk_unpack_reconfig_data_format_srca_impl_ for rationale.
```
Comment (ncvetkovicTT) [inline tt_metal/tt-llk/tt_llk_wormhole_b0/llk_lib/llk_unpack_common.h]: Also, if this is a proper fix for multiple cases, then there's no need to tag the issue number, only state why we're doing this here.

**Review by ncvetkovicTT — COMMENTED** [review]
Please make sure we're not breaking other use cases by moving the `_llk_unpack_reconfig_zero_src_flag_` out.

**Review by amahmudTT — APPROVED** [review]

File: tt_metal/tt-llk/tt_llk_wormhole_b0/llk_lib/llk_unpack_common.h (hunk 7 above)
Comment (nstamatovicTT) [reply tt_metal/tt-llk/tt_llk_wormhole_b0/llk_lib/llk_unpack_common.h]: Yes, the fix sets the flag at reconfig time, before any FPU op runs. So the regular L1->SrcReg->FPU->Dest path is covered too. MOVA2D was just the path that exposed it in sort. Updated the comment.

File: tt_metal/tt-llk/tests/sources/sort_test.cpp [hunk 9]
```diff
@@ -0,0 +1,592 @@
...
+constexpr int SORT_END_PHASE = 5; // log2(64)-1
+
+// Sort direction (this LLK reproducer always sorts ascending).
+constexpr bool ASCENDING = true;
+
+// Helper: count log2 of a power-of-2 integer.
+static constexpr std::uint32_t ilog2_ce(std::uint32_t n)
+{
+    std::uint32_t r = 0;
+    while ((1u << r) < n)
```
Comment (fvranicTT) [inline tt_metal/tt-llk/tests/sources/sort_test.cpp:47]: There's a builtin we could use:

```cpp
static constexpr std::uint32_t ilog2_ce(std::uint32_t n)
{
    return n <= 1u ? 0u : 32u - static_cast<std::uint32_t>(__builtin_clz(n - 1));
}
```

**Review by ncvetkovicTT — APPROVED** [review]

File: tt_metal/tt-llk/tests/sources/sort_test.cpp (hunk 9 above)
Comment (nstamatovicTT) [reply tt_metal/tt-llk/tests/sources/sort_test.cpp:47]: We can't use `__builtin_clz` since zbb(clz instruction) is not supported on WH.
